pydantic/_internal/_generate_schema.py

```diff
@@ -2425,12 +2392,108 @@ def _common_field(
     }
 
 
+class CollectedInvalid(Exception):
+    pass
+
+
+_NOT_INLINABLE_META_KEYS: set[str] = {
+    *CoreMetadata.__annotations__.keys(),
+    _discriminated_union.CORE_SCHEMA_METADATA_DISCRIMINATOR_PLACEHOLDER_KEY,
+    TAGGED_UNION_TAG_KEY,
+}
+
+
 class _Definitions:
     """Keeps track of references and definitions."""
 
     def __init__(self) -> None:
-        self.seen: set[str] = set()
-        self.definitions: dict[str, core_schema.CoreSchema] = {}
+        self.used_deferred_discriminators = False
+        self._recursively_seen: set[str] = set()
+        self._definitions: dict[str, CoreSchema] = {}
+        # Def schema instances of another model, not to be exposed via get_def
+        self._unpacked_definitions: dict[str, CoreSchema] = {}
+
+    def get_def(self, ref: str) -> CoreSchema | None:
+        return self._definitions.get(ref)
+
+    def reference_schema(self, schema: CoreSchema) -> core_schema.DefinitionReferenceSchema:
+        ref = schema['ref']  # type: ignore
+        self._definitions[ref] = schema
+        return core_schema.definition_reference_schema(ref)
```

Comment:
Wouldn't it be better to just pass `ref` here as well, as we only call this when we know `ref` is present and have confirmed that? Removes the need for the `# type: ignore`

Reply:
I don't love this new function. This pattern happens a lot, so I'm not opposed to coupling the last two lines, but I think we need to name it something different, because the name only implies the logic of the last line

Reply:
`create_reference_to_schema`? 🤔 

> Wouldn't it be better to just pass ref here as well

Well maybe, but I would still have an assertion to check that there wont be coming anything corrupt when ref is coming from arg. So: `assert schema['ref'] == ref`